### ISSUE TITLE
Guidance required : generate boot.img and boot.sign using yocto project for secure boot

### ISSUE THREAD

**Issue by Vishwasrao1** [agherzan/meta-raspberrypi#1071]
As of now , I was manually making boot.img, signing it using usbboot(https://github.com/raspberrypi/usbboot) tool for my custom yocto based image.

I would like to incorporate above step in yocto toolchain. So that I will get final signed image from yocto itself. The hardware is rpi CM4. Once I have achieved this I would like to enhance this and generate complete chain of trust including rootfs with help of yocto project.

I was not sure how to start with this, I was wondering if someone from the community has already given it a try or know how to proceed with this. If anyone can guide me or would like to contribute that will be really helpful.

**Comment by agherzan** [comment]
This looks like something useful to have in. We would need to have a way to generate the images that create a signed fat fs image. It shouldn't be that complicated, to be honest. We would need to move things from the boot partition to a filesystem image file with the ability to expose the key for signing. Obviously, the integration of the RPI sign tools will also be needed. This sounds like a variable which would enable the above. How does that sound to you?

**Comment by Vishwasrao1** [comment]
Thank You @kraj for your comment.
@agherzan Yes, you understood it correctly  and that's what I am currently working on.

Right now I am trying to achieve it using post-build script in yocto. People have already used postbuild.sh in buildroot to achieve the same. (https://github.com/raspberrypi/usbboot/issues/147)
So my goal is: 
1. copy content of boot partition to specific path. 
2. Use this content to generate the boot.img utilising usbboot make-boot-image.   
3. Sign this image using usbboot rpi-eeprom-digest.
4. In the end copy this newly generated boot.img and boot.sign into original boot partition 
5. Make the image using this updated boot partition.

I will use the same key (used in point 3) to sign the EEPROM OTP Fuses. Doing this we can achieve a secure root of trust with signed boot and kernel image. I am yet to figure out about signing the rootFS and adding it to the chain of trust. This is something I will do next.

So any inputs regarding post build script or if you know any alternative in yocto that will be also helpful.

**Comment by kraj** [comment]
my suggestion is to bundle the final artifacts using wic process. Having custom SD card generation scripts or process should be avoided for best results.

**Comment by agherzan** [comment]
@kraj of course. I didn't propose external tooling to wic.

The way I see it right now is the following. When `SIGN = 1` (variable name to be defined):

1. Filter out assets that are meant to be deployed as a signed fat image from `IMAGE_BOOT_FILES`.
2. Have `wic` depend on a `recipe:do_deploy` (which depends on the recipes providing/deploying all the assets needed in the signed image). This new recipe will generate the signed image by taking the input from the deploy directory and generating the signed image based on a key defined in another variable.
3. Add the signed file in `IMAGE_BOOT_FILES`.

**Comment by agherzan** [comment]
As a side note, 1 above can be optional but I don't see why not do this clean and avoid duplication (that has an impact on handling updates for example).

**Comment by BuriKizilkaya** [comment]
What is the state of this issue?

**Comment by berkutta** [comment]
Anyone found a good way to setup a hook to read `IMAGE_BOOT_FILES` after it is populated and also logic to iterate through `IMAGE_BOOT_FILES`?

**Comment by lhoward** [comment]
I've been looking into this a bit. I haven't had a chance to test, but if you're using Mender I think it should be possible to use the bootimg it creates, and just sign that. According to Claude, there are a few things missing (`DEPLOYPAYLOAD`, `FATPAYLOAD`, `image-version-info` stamp) but I don't think these are necessary unless you are using those features.

Note also that using the same key to sign `boot.img` as the U-Boot fitImage is not necessarily a good idea; I've done it here for convenience.

```
inherit image_types

BOOT_IMG_PRIVATE_KEY_PEM ?= "${UBOOT_SIGN_KEYDIR}/${UBOOT_SIGN_KEYNAME}.key"

do_image_signed_bootimg[depends] += " \
    rpi-eeprom-native:do_populate_sysroot \
    openssl-native:do_populate_sysroot \
    xxd-native:do_populate_sysroot \
"

IMAGE_CMD:signed-bootimg() {
    if [ -z "${BOOT_IMG_PRIVATE_KEY_PEM}" ]; then
        bbfatal "BOOT_IMG_PRIVATE_KEY_PEM is not set. Please specify the path to the private key .pem file."
    fi

    if [ ! -f "${BOOT_IMG_PRIVATE_KEY_PEM}" ]; then
        bbfatal "Private key not found at ${BOOT_IMG_PRIVATE_KEY_PEM}"
    fi

    BOOTIMG="${IMGDEPLOYDIR}/${IMAGE_LINK_NAME}.bootimg"
    BOOTSIG="${IMGDEPLOYDIR}/${IMAGE_NAME}.bootimg.sig"

    # Sign using the upstream RPi tool for compatibility with the
    # VideoCore GPU bootloader secure boot verification
    rpi-eeprom-digest -k "${BOOT_IMG_PRIVATE_KEY_PEM}" -i "${BOOTIMG}" -o "${BOOTSIG}"
}

IMAGE_TYPEDEP:signed-bootimg:append = " bootimg"
```

**Comment by berkutta** [comment]
For anyone implementing boot.img in Yocto this might be interesting: https://softwarecraft.ch/secure-boot-yocto/

**Comment by lhoward** [comment]
> For anyone implementing boot.img in Yocto this might be interesting: https://softwarecraft.ch/secure-boot-yocto/

Yes, that was actually my starting point until I realized that Mender did a lot of that already. Of course, not everyone is using Mender.

The above link could be improved by making it an image class, and extracting the VFAT image from the sdimg rather than recreating it.
